gwctl/pkg/resourcediscovery/discoverer.go

```diff
@@ -284,19 +303,56 @@ func (d Discoverer) discoverHTTPRoutesFromBackends(ctx context.Context, resource
 	}
 
 	for _, httpRoute := range httpRoutes {
-		var found bool
+		// An HTTPRoute will be included in the resourceModel if it references some
+		// Backend which already exists in the resourceModel.
+		var includeRouteInResourceModel bool
+
 		for _, backendRef := range relations.FindBackendRefsForHTTPRoute(httpRoute) {
+			// Check if the referenced backend exists in the resourceModel.
 			backendID := BackendID(backendRef.Group, backendRef.Kind, backendRef.Namespace, backendRef.Name)
-			_, ok := resourceModel.Backends[backendID]
+			backendNode, ok := resourceModel.Backends[backendID]
 			if !ok {
 				continue
 			}
-			found = true
+
+			// Ensure that if this is a cross namespace reference, then it is accepted
+			// through some ReferenceGrant.
+			if httpRoute.GetNamespace() != backendRef.Namespace {
```

Comment:
Not a blocker for this PR, but I think it would be very helpful to have a more generic RefGrant library that both gwctl and Gateway implementations can use. This has some of the foundational bits for this, but it would just need to be structured differently if that were a goal.

Reply:
Nice idea. I'll try to come up with something more generic separately and will bring it up for discussion.